### PR TITLE
pyfa: 2.60.2 -> 2.61.0

### PR DESCRIPTION
**Pull request by paschoal**
switch from appImage to pythonApplication, update from 2.60.2 to 2.61.0
Pyfa Release [2.61.0](https://github.com/pyfa-org/Pyfa/releases/tag/v2.61.0) 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by ToasterUwU — CHANGES_REQUESTED** [review]
Im getting the following error when trying to build it. It also eats like a solid 12GB of RAM, not sure if that's expected.

`error: evaluation aborted with the following error message: 'lib.customisation.callPackageWith: Function called without required argument "wrapGAppsHook" at /home/aki/.cache/nixpkgs-review/pr-356622-2/nixpkgs/pkgs/by-name/py/pyfa/package.nix:7, did you mean "wrapGAppsHook3" or "wrapGAppsHook4"?'`

**Comment by paschoal** [comment]
> Im getting the following error when trying to build it. It also eats like a solid 12GB of RAM, not sure if that's expected.
> 
> `error: evaluation aborted with the following error message: 'lib.customisation.callPackageWith: Function called without required argument "wrapGAppsHook" at /home/aki/.cache/nixpkgs-review/pr-356622-2/nixpkgs/pkgs/by-name/py/pyfa/package.nix:7, did you mean "wrapGAppsHook3" or "wrapGAppsHook4"?'`

i'm not sure if I understand why in my local env I was able to build with wrapGAppsHook only, but will take a look at this later. 

I just switched to version 4, but I wonder if would be better to validate the local version available and decide which to use - what is your take?.

in regards to memory, I'm all ear on how to optimize this, the pyinstaller method is taken from [pyfa wiki](https://github.com/pyfa-org/Pyfa/wiki/Developers:-Creating-Executable-Builds-for-Pyfa)

**Comment by paschoal** [comment]
> I just switched to version 4, but I wonder if would be better to validate the local version available and decide which to use - what is your take?.

made the wrong choice I think, pyfa is gtk3 -- switched to it.

**Comment by ToasterUwU** [comment]
I'm unsure how other Python based packages handle this, so I guess that's something to look into.
Theoretically there is no need to turn this into a proper executable, it would also be possible to just run it via python through a wrapper script that takes the place of the binary.
But I'm unsure what's the norm and best practice on Nixpkgs.

I'm going to bed now, if I find time tommorow, I will look into it.

**Comment by paschoal** [comment]
> I'm unsure how other Python based packages handle this, so I guess that's something to look into. Theoretically there is no need to turn this into a proper executable, it would also be possible to just run it via python through a wrapper script that takes the place of the binary. But I'm unsure what's the norm and best practice on Nixpkgs.
> 
> I'm going to bed now, if I find time tommorow, I will look into it.

yeap, while I do see the benefit of compiling it and making sure that all the libraries are properly set, let me see if we can just wrap it and run.

**Comment by paschoal** [comment]
> let me see if we can just wrap it and run.

to no avail, getting all sorts of error and missing dependencies, and can't just install it directly, since it is an externally managed environment.

let me know if you have any luck.

meanwhile did you manage to run it locally using pyinstaller version? the wrapGAppsHook was solved?

**Comment by ToasterUwU** [comment]
i just got around to test building this, and got an error.

Btw, the reason your GApps thing didnt work is because you arent using the unstable branch. At least im pretty sure you are making and testing this on stable NixOS, but when this gets merged and build, it does so on unstable, so you should use that for making PRs here.

```
error:
       … while querying the derivation named 'pyfa-2.61.0'

       … while evaluating the attribute 'out.outPath'
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/lib/customisation.nix:352:13:
          351|             drvPath = assert condition; drv.${outputName}.drvPath;
          352|             outPath = assert condition; drv.${outputName}.outPath;
             |             ^
          353|           } //

       … while calling the 'getAttr' builtin
         at <nix/derivation-internal.nix>:44:19:
           43|       value = commonAttrs // {
           44|         outPath = builtins.getAttr outputName strict;
             |                   ^
           45|         drvPath = strict.drvPath;

       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:34:12:
           33|
           34|   strict = derivationStrict drvAttrs;
             |            ^
           35|

       … while evaluating derivation 'pyfa-2.61.0'
         whose name attribute is located at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'propagatedBuildInputs' of derivation 'pyfa-2.61.0'
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:390:7:
          389|       depsHostHostPropagated      = elemAt (elemAt propagatedDependencies 1) 0;
          390|       propagatedBuildInputs       = elemAt (elemAt propagatedDependencies 1) 1;
             |       ^
          391|       depsTargetTargetPropagated  = elemAt (elemAt propagatedDependencies 2) 0;

       … while calling anonymous lambda
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:326:13:
          325|       (map (drv: getDev drv.__spliced.hostHost or drv) (checkDependencyList "depsHostHostPropagated" depsHostHostPropagated))
          326|       (map (drv: getDev drv.__spliced.hostTarget or drv) (checkDependencyList "propagatedBuildInputs" propagatedBuildInputs))
             |             ^
          327|     ]

       … from call site
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:326:18:
          325|       (map (drv: getDev drv.__spliced.hostHost or drv) (checkDependencyList "depsHostHostPropagated" depsHostHostPropagated))
          326|       (map (drv: getDev drv.__spliced.hostTarget or drv) (checkDependencyList "propagatedBuildInputs" propagatedBuildInputs))
             |                  ^
          327|     ]

       … while calling 'getOutput'
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/lib/attrsets.nix:1796:23:
         1795|   */
         1796|   getOutput = output: pkg:
             |                       ^
         1797|     if ! pkg ? outputSpecified || ! pkg.outputSpecified

       … while evaluating a branch condition
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/lib/attrsets.nix:1797:5:
         1796|   getOutput = output: pkg:
         1797|     if ! pkg ? outputSpecified || ! pkg.outputSpecified
             |     ^
         1798|       then pkg.${output} or pkg.out or pkg

       … in the left operand of the OR (||) operator
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/lib/attrsets.nix:1797:32:
         1796|   getOutput = output: pkg:
         1797|     if ! pkg ? outputSpecified || ! pkg.outputSpecified
             |                                ^
         1798|       then pkg.${output} or pkg.out or pkg

       … in the argument of the not operator
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/lib/attrsets.nix:1797:10:
         1796|   getOutput = output: pkg:
         1797|     if ! pkg ? outputSpecified || ! pkg.outputSpecified
             |          ^
         1798|       then pkg.${output} or pkg.out or pkg

       … from call site
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:326:25:
          325|       (map (drv: getDev drv.__spliced.hostHost or drv) (checkDependencyList "depsHostHostPropagated" depsHostHostPropagated))
          326|       (map (drv: getDev drv.__spliced.hostTarget or drv) (checkDependencyList "propagatedBuildInputs" propagatedBuildInputs))
             |                         ^
          327|     ]

       … while calling anonymous lambda
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/lib/lists.nix:334:29:
          333|   */
          334|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                             ^
          335|

       … from call site
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/lib/lists.nix:334:32:
          333|   */
          334|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                                ^
          335|

       … while calling anonymous lambda
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:281:15:
          280|     imap1
          281|       (index: dep:
             |               ^
          282|         if isDerivation dep || dep == null || builtins.isString dep || builtins.isPath dep then dep

       … while evaluating a branch condition
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:282:9:
          281|       (index: dep:
          282|         if isDerivation dep || dep == null || builtins.isString dep || builtins.isPath dep then dep
             |         ^
          283|         else if isList dep then checkDependencyList' ([index] ++ positions) name dep

       … in the left operand of the OR (||) operator
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:282:69:
          281|       (index: dep:
          282|         if isDerivation dep || dep == null || builtins.isString dep || builtins.isPath dep then dep
             |                                                                     ^
          283|         else if isList dep then checkDependencyList' ([index] ++ positions) name dep

       … in the left operand of the OR (||) operator
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:282:44:
          281|       (index: dep:
          282|         if isDerivation dep || dep == null || builtins.isString dep || builtins.isPath dep then dep
             |                                            ^
          283|         else if isList dep then checkDependencyList' ([index] ++ positions) name dep

       … in the left operand of the OR (||) operator
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:282:29:
          281|       (index: dep:
          282|         if isDerivation dep || dep == null || builtins.isString dep || builtins.isPath dep then dep
             |                             ^
          283|         else if isList dep then checkDependencyList' ([index] ++ positions) name dep

       … from call site
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:282:12:
          281|       (index: dep:
          282|         if isDerivation dep || dep == null || builtins.isString dep || builtins.isPath dep then dep
             |            ^
          283|         else if isList dep then checkDependencyList' ([index] ++ positions) name dep

       … while calling 'isDerivation'
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/lib/attrsets.nix:1282:5:
         1281|   isDerivation =
         1282|     value: value.type or null == "derivation";
             |     ^
         1283|

       … while calling the 'elemAt' builtin
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/lib/lists.nix:334:43:
          333|   */
          334|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                                           ^
          335|

       … while calling 'checkDrv'
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/development/interpreters/python/mk-python-derivation.nix:206:26:
          205|
          206|     checkDrv = attrName: drv:
             |                          ^
          207|       if (isPythonModule drv) && (isMismatchedPython drv) then throwMismatch attrName drv

       … while evaluating a branch condition
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/development/interpreters/python/mk-python-derivation.nix:207:7:
          206|     checkDrv = attrName: drv:
          207|       if (isPythonModule drv) && (isMismatchedPython drv) then throwMismatch attrName drv
             |       ^
          208|       else drv;

       … in the left operand of the AND (&&) operator
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/development/interpreters/python/mk-python-derivation.nix:207:31:
          206|     checkDrv = attrName: drv:
          207|       if (isPythonModule drv) && (isMismatchedPython drv) then throwMismatch attrName drv
             |                               ^
          208|       else drv;

       … from call site
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/development/interpreters/python/mk-python-derivation.nix:207:11:
          206|     checkDrv = attrName: drv:
          207|       if (isPythonModule drv) && (isMismatchedPython drv) then throwMismatch attrName drv
             |           ^
          208|       else drv;

       … while calling 'isPythonModule'
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/development/interpreters/python/mk-python-derivation.nix:42:20:
           41|
           42|   isPythonModule = drv:
             |                    ^
           43|     # all pythonModules have the pythonModule attribute

       … in the left operand of the AND (&&) operator
         at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/development/interpreters/python/mk-python-derivation.nix:46:5:
           45|     # Some pythonModules are turned in to a pythonApplication by setting the field to false
           46|     && (!isBool drv.pythonModule);
             |     ^
           47|

       error: undefined variable 'dateutil'
       at /home/aki/.cache/nixpkgs-review/pr-356622/nixpkgs/pkgs/by-name/py/pyfa/package.nix:30:5:
           29|     matplotlib
           30|     dateutil
             |     ^
```

**Comment by paschoal** [comment]
> i just got around to test building this, and got an error.
> 
> Btw, the reason your GApps thing didnt work is because you arent using the unstable branch. At least im pretty sure you are making and testing this on stable NixOS, but when this gets merged and build, it does so on unstable, so you should use that for making PRs here.
> 

Oh, so while my system is running 100% unstable, the branch that I trigger the build process on nixos fork locally was stable, ha! did not realized that.

probably also related to this other error you have, gonna switch the branch and test it out later.

**Comment by paschoal** [comment]
oddly, rebased the branch against nixos-unstable and built it again without errors, I can't reproduce the errors you got locally.



got any other clue of what my be going on?

**Comment by paschoal** [comment]
manually did a check on the subpackages that was included from python3Packages (python312Packages) and the dateutil was indeed incorrect, adjusted to python-dateutil.

so that error you had of ```undefined variable 'dateutil'``` should be gone, but the question on how we make this reproducible behaviours is still open.

**Comment by paschoal** [comment]
and then, just did the nixpkgs-review step (hopefully that is the answer to my previous question about making it reproducible?)

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -16724,6 +16724,12 @@
     githubId = 33826198;
     name = "Philipp Arras";
   };
+  paschoal = {
+    email = "paschoal@gmail.com";
+    github = "paschoal";
+    githubId = 64602;
+    name = "Matheus Paschoal";
+  };
```
Comment (pluiedev) [inline maintainers/maintainer-list.nix:16732]: Should be in its own commit

File: pkgs/by-name/py/pyfa/package.nix [hunk 2]
```diff
@@ -1,28 +1,114 @@
...
+  format = "other";
 
-  src = fetchurl {
-    name = "pyfa-appimage-${version}";
-    url = "https://github.com/pyfa-org/Pyfa/releases/download/v${version}/pyfa-v${version}-linux.AppImage";
-    hash = "sha256-6doetQ6E1Occ/SqewfxRqPEX1MnUuFomm+8VmetIz4Y=";
+  src = fetchFromGitHub {
+    owner = "pyfa-org";
+    repo = "Pyfa";
+    rev = "v${version}";
```
Comment (pluiedev) [inline pkgs/by-name/py/pyfa/package.nix:21]: Prefer prefixing tags with `refs/tags/` to avoid collision with branches
```suggestion
    rev = "refs/tags/v${version}";
```

File: pkgs/by-name/py/pyfa/package.nix [hunk 3]
```diff
@@ -1,28 +1,114 @@
...
+    hash = "sha256-VSuRQugUODc+LbhKbzsA09pnqPEIAt0pQS2An/p7r9A=";
   };
 
+  build-system = [ python3Packages.setuptools ];
+  propagatedBuildInputs = with python3Packages; [
```
Comment (pluiedev) [inline pkgs/by-name/py/pyfa/package.nix:26]: Python dependencies should be `dependencies`
```suggestion
  dependencies = with python3Packages; [
```

File: pkgs/by-name/py/pyfa/package.nix [hunk 4]
```diff
@@ -1,28 +1,114 @@
...
+    beautifulsoup4
+    pyaml
+    roman
+    numpy
+    python-jose
+    requests-cache
+  ];
+
+  buildInputs = [
+    python3Packages.pyinstaller
```
Comment (pluiedev) [inline pkgs/by-name/py/pyfa/package.nix:44]: This sounds more like a nativeBuildInput for me 🤔

File: pkgs/by-name/py/pyfa/package.nix [hunk 5]
```diff
@@ -1,28 +1,114 @@
...
+    cat > setup.py <<EOF
+      from setuptools import setup
+      setup(
+        name = "${pname}",
+        version = "${version}",
+        scripts = ["pyfa.py"],
+        packages = setuptools.find_packages(),
+      )
+    EOF
+  '';
```
Comment (pluiedev) [inline pkgs/by-name/py/pyfa/package.nix:65]: A comment on why this is necessary would be nice (e.g. upstream's setup.py is too complicated, missing setup.py, etc.)

File: pkgs/by-name/py/pyfa/package.nix [hunk 6]
```diff
@@ -1,28 +1,114 @@
...
+
+  configurePhase = ''
+    python db_update.py
+  '';
```
Comment (pluiedev) [inline pkgs/by-name/py/pyfa/package.nix:69]:
```suggestion
  configurePhase = ''
  	runHook preConfigure
  	
    python3 db_update.py
    
  	runHook postConfigure
  '';
```

File: pkgs/by-name/py/pyfa/package.nix [hunk 7]
```diff
@@ -1,28 +1,114 @@
...
+
+  buildPhase = ''
+    ${python3Packages.pyinstaller}/bin/pyinstaller \
+      --clean \
+      --noconfirm \
+      pyfa.spec
+  '';
```
Comment (pluiedev) [inline pkgs/by-name/py/pyfa/package.nix:76]: Fairly sure `pyinstaller` would already be in PATH

```suggestion
  buildPhase = ''
  	runHook preBuild
  	
    pyinstaller --clean --noconfirm pyfa.spec
    
    runHook postBuild
  '';
```

File: pkgs/by-name/py/pyfa/package.nix [hunk 8]
```diff
@@ -1,28 +1,114 @@
...
+  # creating a symbolic link out in $out/bin to $out/share/pyfa to avoid
+  # exposing the innards of pyfa to the rest of the env.
+  #
+  installPhase = ''
+    mkdir -p $out/bin
+    mkdir -p $out/share/pixmaps
+
+    cp -r dist/pyfa $out/share/
+    cp imgs/gui/pyfa64.png $out/share/pixmaps/pyfa.png
+    ln -sf $out/share/pyfa/pyfa $out/bin/pyfa
```
Comment (pluiedev) [inline pkgs/by-name/py/pyfa/package.nix:90]: You can just run `wrapProgram` below and skip this

File: pkgs/by-name/py/pyfa/package.nix [hunk 9]
```diff
@@ -1,28 +1,114 @@
...
+  '';
+
+  fixupPhase = ''
+    runHook preFixup
+
+    wrapProgramShell $out/share/pyfa/pyfa \
+      ''${gappsWrapperArgs[@]} \
+
+    runHook postFixup
+  '';
```
Comment (pluiedev) [inline pkgs/by-name/py/pyfa/package.nix:100]:
```suggestion
  preFixup = ''
    makeWrapperArgs+=(''${gappsWrapperArgs[@]})
  '';
```

File: pkgs/by-name/py/pyfa/package.nix [hunk 10]
```diff
@@ -1,28 +1,114 @@
...
+
+  doCheck = false;
```
Comment (pluiedev) [inline pkgs/by-name/py/pyfa/package.nix:102]: Leave a comment explaining why

File: pkgs/by-name/py/pyfa/package.nix [hunk 11]
```diff
@@ -1,28 +1,114 @@
...
     description = "Python fitting assistant, cross-platform fitting tool for EVE Online";
     homepage = "https://github.com/pyfa-org/Pyfa";
     license = lib.licenses.gpl3Plus;
     maintainers = with lib.maintainers; [
       toasteruwu
       cholli
+      paschoal
     ];
     mainProgram = "pyfa";
     sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
```
Comment (pluiedev) [inline pkgs/by-name/py/pyfa/package.nix:114]: Remove if building from source
```suggestion
```

**Review by pluiedev — CHANGES_REQUESTED** [review]

File: pkgs/by-name/py/pyfa/package.nix (hunk 9 above)
Comment (paschoal) [reply pkgs/by-name/py/pyfa/package.nix:100]: not sure if this change is correct, already attempted it before and it does render the following error
```
(pyfa:33501): GLib-GIO-ERROR **: 14:14:11.871: No GSettings schemas are installed on the system
zsh: trace trap (core dumped)  ./bin/pyfa
```
and break the application.

while using the fixupPhase strategy does solve problem.
maybe something else in the preFixup result would solve the problem?

Comment (pluiedev) [reply pkgs/by-name/py/pyfa/package.nix:100]: It'd work if a) the file is executable and b) has a python shebang. Lemme try this for myself first

**Comment by paschoal** [comment]
```
[:/data/git/nixpkgs] update-pyfa-2.61.0 ± nix-shell -p nixfmt-rfc-style --run "nixfmt --check 'pkgs/by-name/py/pyfa/package.nix'"
[:/data/git/nixpkgs] update-pyfa-2.61.0 ± nix-shell -p nixfmt-classic --run "nixfmt --check 'pkgs/by-name/py/pyfa/package.nix'"
pkgs/by-name/py/pyfa/package.nix: not formatted
```
:confused:

**Comment by gepbird** [comment]
> ```
> [:/data/git/nixpkgs] update-pyfa-2.61.0 ± nix-shell -p nixfmt-rfc-style --run "nixfmt --check 'pkgs/by-name/py/pyfa/package.nix'"
> [:/data/git/nixpkgs] update-pyfa-2.61.0 ± nix-shell -p nixfmt-classic --run "nixfmt --check 'pkgs/by-name/py/pyfa/package.nix'"
> pkgs/by-name/py/pyfa/package.nix: not formatted
> ```
> 
> 😕

You can format it with `nixfmt-rfc-style` (omit the `--check`) and commit the changes: `nix-shell -p nixfmt-rfc-style --run "nixfmt 'pkgs/by-name/py/pyfa/package.nix'"`.

Nit: can you configure your editor to use 2 space indentation in nixpkgs? There are tabs in `configurePhase` and likely that's why you didn't spot the formatting issue.

**Comment by paschoal** [comment]
> Nit: can you configure your editor to use 2 space indentation in nixpkgs? There are tabs in `configurePhase` and likely that's why you didn't spot the formatting issue.

interesting, the code before the review was nixfmt-rfc-style complaint, and my local configuration do use 2 space width; but the last change commited directly from the suggestions [9eba09](https://github.com/NixOS/nixpkgs/pull/356622/commits/9eba0998f0a9199ee235443a95f312d9a44b7877) was the one.

my fault was not rebasing to the very last commit that do introduce the changes, I was under the impression that [cc4d769](https://github.com/NixOS/nixpkgs/pull/356622/commits/cc4d7695e85e9b5f284f8b253ea80bfbed973e8c) broke it, but it was [9eba09](https://github.com/NixOS/nixpkgs/pull/356622/commits/9eba0998f0a9199ee235443a95f312d9a44b7877). Rebasing with the github changes do identify the issue using ```--check```.

**Comment by gepbird** [comment]
Oh I'm just seeing that the editorconfig check is failing, and nixfmt doesn't fix that. You fixed some whitespace issues but there are 2 tabs which it doesn't like. To convert tabs to 2 spaces you can run `sed -i 's/\t/  /g' pkgs/by-name/py/pyfa/package.nix` and then you should run the nixfmt command as before. I'm happy to help with configuring this in your editor to not have these issues in the future, I just don't know what you're using :)

**Comment by ToasterUwU** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 356622`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>pyfa</li>
  </ul>
</details>

**Comment by ToasterUwU** [comment]
@paschoal At first glance, im happy with the current state of this. Please make sure you force push commits into this branch that follow the guidelines. Meaning one commit for adding yourself to the maintainer file, and one commit for updating the package.

You can find the naming scheme for those commits in lots of PRs, including the ones I was involved in. On the run rn, so I cant find it for you, sorry.

**Comment by Daholli** [comment]
The easiest way of doing this is probably a `git reset --soft 8edf06bea5bcbee082df1b7369ff973b91618b8d` please double-check that this is the commit before your first commit

Then commit the `maintainer add `, and then the `pyfa: 2.61.2 -> 2.62.0`

Please double-check all of what I wrote, I am a bit feverish and am typing this on the phone.

**Review by Daholli — CHANGES_REQUESTED** [review]
I am all happy after the commits are fixed

Also sorry for making you do all of this, since we are both busy

**Review by Daholli — APPROVED** [review]

**Comment by Daholli** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 356622`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>pyfa</li>
  </ul>
</details>

**Comment by Daholli** [comment]
Yeah works flawless, tried esi auth, importing exporting, creating new fittings.

Lets get this merged :D

**Review by ToasterUwU — APPROVED** [review]

**Review by gepbird — APPROVED** [review]

**Review by Daholli — APPROVED** [review]

**Review by ToasterUwU — APPROVED** [review]

File: pkgs/by-name/py/pyfa/package.nix [hunk 12]
```diff
@@ -1,31 +1,124 @@
...
+  nativeBuildInputs = [
+    python3Packages.pyinstaller
+    wrapGAppsHook3
+  ];
+
+  #
+  # upstream does not include setup.py
+  #
+  patchPhase = ''
+    cat > setup.py <<EOF
```
Comment (pbsds) [inline pkgs/by-name/py/pyfa/package.nix:59]:
```suggestion
    cat > setup.py <<"EOF"
```

Comment (Daholli) [reply pkgs/by-name/py/pyfa/package.nix:59]: It is currently how it was in the documentation I found regarding no setup.py are you sure this would work? 

I thought the EOF token needs to be by itself

Comment (paschoal) [reply pkgs/by-name/py/pyfa/package.nix:59]: this would enter the realm of bash nuance operations

[bash.html#Here-Documents](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Here-Documents
)
> No parameter and variable expansion, command substitution, arithmetic expansion, or filename expansion is performed on word. If any part of word is quoted, the delimiter is the result of quote removal on word, and the lines in the here-document are not expanded. If word is unquoted, all lines of the here-document are subjected to parameter expansion, command substitution, and arithmetic expansion, the character sequence \newline is ignored, and ‘\’ must be used to quote the characters ‘\’, ‘$’, and ‘`’.

sorry, I fail to see the benefit of limiting operations in this case since the unquoted is more trivial / widely used and in this case both yield the same result, can you please detail more situations where using the quoted heredoc would come handy?

**Comment by paschoal** [comment]
rebase against master and push to solve conflict @ maintainers/maintainers-list.nix

File: pkgs/by-name/py/pyfa/package.nix (hunk 12 above)
Comment (pbsds) [reply pkgs/by-name/py/pyfa/package.nix:59]: Removing footguns is clean practice. The fact that here-documents are not commonly quoted in nixpkgs doesn't mean that that is how it should be.

File: pkgs/by-name/py/pyfa/package.nix [hunk 13]
```diff
@@ -1,31 +1,124 @@
...
+      from setuptools import setup
+      setup(
+        name = "${pname}",
+        version = "${version}",
```
Comment (pbsds) [inline pkgs/by-name/py/pyfa/package.nix:63]: instead of escaping `EOF` you could
```suggestion
        name = "$pname",
        version = "$version",
```
in turn making the version overrideable

File: pkgs/by-name/py/pyfa/package.nix (hunk 12 above)
Comment (paschoal) [reply pkgs/by-name/py/pyfa/package.nix:59]: Sure, but sorry, gonna have to insist because I still don't see the benefit or find any documentation that justify the clean practice with a good reasoning, can you recomend me any good writing on that?

File: pkgs/by-name/py/pyfa/package.nix (hunk 13 above)
Comment (paschoal) [reply pkgs/by-name/py/pyfa/package.nix:63]: searching nixos/nixpkgs for $pname declaration without braces, I don't see any adoption of this practice;

maybe a good write up on it would help understanding the benefit and increasing the adoption of the pratice?

**Review by ToasterUwU — APPROVED** [review]

**Review by Daholli — APPROVED** [review]

File: pkgs/by-name/py/pyfa/package.nix (hunk 12 above)
Comment (pluiedev) [reply pkgs/by-name/py/pyfa/package.nix:59]: My 2¢ is to follow whatever ShellCheck recommends, and it seems to be completely happy with unquoted heredocs:

image

File: pkgs/by-name/py/pyfa/package.nix (hunk 13 above)
Comment (pluiedev) [reply pkgs/by-name/py/pyfa/package.nix:63]: In Nix derivations, attributes are automatically passed to the builder script as Bash variables. You can try running echo $pname yourself and see that it exists and it is set to the current value of pname.

The main benefit of this theoretically is that when a package is overridden with overrideAttrs, the setup.py reacts to the override as it is synchronized with the overridden variables. Using string substitution will bake the values directly into the builder, meaning that an override won't change the generated setup.py.

*That said,* I'm not 100% certain that this will cause a rebuild when pname and version are overridden, unlike using other override-aware mechanisms like `finalAttrs`. Try if it works first

Comment (pbsds) [reply pkgs/by-name/py/pyfa/package.nix:63]: https://matt-rickard.com/avoiding-footguns

If you want a silly example:

> [!CAUTION]
> DO NOT RUN THIS

```shell
cat <<EOF >foobar.py
# prevent use of `sudo rm -rf / --no-preserve-root`
with open(file, "r") as f:
    data = f.read()
data = data.replace(
    "sudo rm -rf / --no-preserve-root",
    "echo sudo rm -rf / --no-preserve-root",
)
with open(file, "w") as f:
    f.write(data)
EOF
```


Avoiding footguns by default should never be considered bad practice. The code might be fine now, but a later contributor may trigger an unwanted case

> searching nixos/nixpkgs for $pname declaration without braces, I don't see any adoption of this practice

true, we actively try and avoid using `pname`.

```suggestion
        name = "pyfa",
        version = "$version",
```

but personally still prefer using `'EOF'`

> Try if it works first

`buildPythonApplication` does not yet support the `finalAttrs` pattern AFAIK.
